subprojects/base-services/src/main/java/org/gradle/internal/hash/HashCode.java

```diff
@@ -88,81 +107,124 @@ private static int decode(char ch) {
         throw new IllegalArgumentException("Illegal hexadecimal character: " + ch);
     }
 
-    public int length() {
-        return bytes.length;
+    @Override
+    public int compareTo(HashCode that) {
+        return Integer.compare(hashCode, that.hashCode);
     }
 
-    public byte[] toByteArray() {
-        return bytes.clone();
-    }
+    @Override
+    public abstract boolean equals(@Nullable Object obj);
 
     @Override
-    public int hashCode() {
-        if (hashCode == 0) {
-            hashCode = (bytes[0] & 0xFF)
-                | ((bytes[1] & 0xFF) << 8)
-                | ((bytes[2] & 0xFF) << 16)
-                | ((bytes[3] & 0xFF) << 24)
-                // Make sure it's always > 0 but without affecting the lower 32 bits
-                | (1L << 32);
+    public abstract int hashCode();
+
+    public abstract int length();
+
+    public abstract byte[] toByteArray();
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder(2 * length());
+        for (byte b : toByteArray()) {
+            sb.append(HEX_DIGITS[(b >> 4) & 0xf])
+                .append(HEX_DIGITS[b & 0xf]);
+        }
+        return sb.toString();
+    }
+
+    static byte[] longsToBytes(long... values) {
+        ByteBuffer buffer = ByteBuffer.allocate(values.length * Long.BYTES);
+        for (long value : values) {
+            buffer.putLong(value);
+        }
+        return buffer.array();
+    }
+
+    static long[] bytesToLongs(byte... values) {
+        int size = (int) Math.ceil((double) values.length / Long.SIZE);
+        LongBuffer buffer = LongBuffer.allocate(size);
+        for (byte b : values) {
+            buffer.put(b);
         }
-        return (int) hashCode;
+        return buffer.array();
+    }
+}
+
+class HashCode16 extends HashCode {
+    private final long long0;
+    private final long long1;
+
+    HashCode16(int hashCode, byte[] values) {
+        this(
+            hashCode,
+            Longs.fromBytes(values[0], values[1], values[2], values[3], values[4], values[5], values[6], values[7]),
+            Longs.fromBytes(values[8], values[9], values[10], values[11], values[12], values[13], values[14], values[15])
+        );
+    }
+
+    HashCode16(int hashCode, long... values) {
+        super(hashCode);
+        this.long0 = values[0];
+        this.long1 = values[1];
+    }
+
+    @Override
+    public int length() {
+        return 2 * Long.BYTES;
     }
 
     @Override
     public boolean equals(@Nullable Object obj) {
         if (obj == this) {
             return true;
-        }
-
-        if (obj == null || obj.getClass() != HashCode.class) {
+        } else if (obj == null || obj.getClass() != HashCode16.class) {
             return false;
         }
+        HashCode16 that = (HashCode16) obj;
+        return long0 == that.long0
```

Comment:
* When the project is up-to-date, HashCodes will mostly be equal.
Sometimes after adding them to Sets, other times they're compared directly (e.g. in other Class's equals methods).
* It's easier/faster to compare two longs, than 2*8 bytes.